README.md
[ci-img]: https://travis-ci.org/ciena-frost/eslint-config-frost-standard.svg "Travis CI Build Status"
[ci-url]: https://travis-ci.org/ciena-frost/eslint-config-frost-standard

[cov-img]: https://coveralls.io/repos/github/ciena-frost/eslint-config-frost-standard/badge.svg?branch=master
[cov-url]: https://coveralls.io/github/ciena-frost/eslint-config-frost-standard

[npm-img]: https://img.shields.io/npm/v/eslint-config-frost-standard.svg "NPM Version"
[npm-url]: https://www.npmjs.com/package/eslint-config-frost-standard

# eslint-config-frost-standard [![ci][ci-img]][ci-url] [![cov][cov-img]][cov-url] [![npm][npm-img]][npm-url]

Standard ESLint rules for Frost projects (based on [`standard`](https://github.com/feross/standard)).
It is our goal to not deviate from `standard` but rather augment it with some additional rules not currently covered
by `standard`. For example, we currently add some rules w.r.t. JSDoc comments.

[![js-standard-style](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)

## Additional Rules

One of the reasons for this package's existence is to be able to define some additional rules that `standard`
didn't see fit to include, but that we find useful/necessary. Here we will highlight what these rules are, but just
in case our documentation ever falls behind :scream:, You can always find the full list of additions [here](index.js).

### `camelcase`
Enforces a `camelCase` variable naming convention. More details available
[here](http://eslint.org/docs/rules/camelcase).

### `complexity`
Enforces a maximum cyclomatic complexity (of 5) for functions. More details available [here](http://eslint.org/docs/rules/complexity).

### `max-len`
Enforces a maximum line length of 120 characters. More details available [here](http://eslint.org/docs/rules/max-len)

### `no-var`
Enforces the use of `let` or `const` instead of `var`. More details available
[here](http://eslint.org/docs/rules/no-var).

### `object-curly-spacing`
Enforces that object literals do not have spaces around the curly braces. More details available
[here](http://eslint.org/docs/rules/object-curly-spacing).

### `no-unused-expressions`
This is mainly to avoid using dangerous property assertions in `chai` like `expect().to.be.null`. Some discussion
about why those are dangerous is available [here](https://github.com/chaijs/chai/issues/726).
More details available [here](http://eslint.org/docs/rules/no-unused-expressions).

### `valid-jsdoc`
Make sure your `JSDoc` comment blocks are valid/consistent. More details available
[here](http://eslint.org/docs/rules/valid-jsdoc).

### `eslint-plugin-ember-standard`
We use a number of rules from
[`eslint-plugin-ember-standard`](https://github.com/ciena-blueplanet/eslint-plugin-ember-standard). More details
about each possible rule are available
[here](https://github.com/ciena-blueplanet/eslint-plugin-ember-standard/tree/master#rules).

#### `ember-standard/computed-property-readonly`
Ensure that all computed properties are read-only.

#### `ember-standard/destructure`
Ensure that the `Ember` object is destructured.

#### `ember-standard/import`
Ensure that the `Ember` is explicitly imported.

#### `ember-standard/logger`
Ensure that the `Ember.Logger` is used instead of `console.log()`.

#### `ember-standard/no-set-in-computed-property`
Ensure that values are not `set()` inside a computed property (no side-effects).

#### `ember-standard/no-settimeout`
Ensure that `Ember.run` is used instead of `setTimeout()`.

#### `ember-standard/prop-types`
Ensure that `PropTypes` references from `ember-prop-types` are valid.

#### `ember-standard/single-destructure`
Ensure Ember properties are destructured in a single variable declaration.

### `eslint-plugin-mocha`
We use a number of rules from [`eslint-plugin-mocha`](https://github.com/lo1tuma/eslint-plugin-mocha)

#### `mocha/handle-done-callback`
Make sure that when you define the `done` callback, you actually call it somewhere. More details available [here](https://github.com/lo1tuma/eslint-plugin-mocha/blob/master/docs/rules/handle-done-callback.md).

#### `mocha/no-exclusive-tests`
Make sure you don't accidentally check in a `describe.only`. More details available [here](https://github.com/lo1tuma/eslint-plugin-mocha/blob/master/docs/rules/no-exclusive-tests.md).

#### `mocha/no-global-tests`
Make sure you don't use an `it()` outside of a `describe()`. More details available [here](https://github.com/lo1tuma/eslint-plugin-mocha/blob/master/docs/rules/no-global-tests.md).

#### `mocha/no-pending-tests`
Make sure you don't include empty `it()` calls. More details available [here](https://github.com/lo1tuma/eslint-plugin-mocha/blob/master/docs/rules/no-pending-tests.md).

#### `mocha/no-skipped-tests`
Emit a warning when you use `it.skip()` or `describe.skip()`. Sometimes it's necessary, but it's nice to easily see them all listed out in one place, and they should be temporary whenever they are checked in. More details available [here](https://github.com/lo1tuma/eslint-plugin-mocha/blob/master/docs/rules/no-skipped-tests.md).

### `eslint-plugin-ocd`
We use a number of rules from [`eslint-plugin-ocd`](https://github.com/ciena-blueplanet/eslint-plugin-ocd).
More details about the rules are available [here](https://github.com/ciena-blueplanet/eslint-plugin-ocd#rules).

#### `ocd/sort-import-declaration-specifiers`
Ensures that named imports within a single import block are sorted alphabetically.

#### `ocd/sort-import-declaration`
Ensures that imports are sorted alphabetically (by the module name, not the thing being imported).

#### `ocd/sort-variable-declarator-properties`
Ensures that destructured variables are sorted alphabetically.


CHANGELOG.md
# 10.1.0 (2019-02-27)

* Upgrade eslint-plugin-ember-standard


# 10.0.1 (2018-09-25)

* Update travis api key


# 10.0.0 (2018-09-25)

* Update eslint-plugin-ember-standard for Ember 3.0


# 9.0.2 (2018-03-28)
* **Updated** engines.node version to `>= 6.0.0` in `package.json`


# 9.0.1 (2018-03-07)
* **Updated** pull request template
* **Added** issue template
* **Updated** `eslint-plugin-ember-standard` to `1.0.3`
* **Updated** `eslint-plugin-ocd` to `1.0.6`
* **Updated** `package-lock.json` file

# 9.0.0 (2017-12-18)
* **Updated** babel-eslint to ^8.0.3
* **Updated** istanbul to ^0.4.5
* **Upgraded** and pinned eslint-plugin-ember-standard to latest version
* **Upgraded** and pinned eslint-plugin-ocd to latest version

# 8.0.0 (2017-11-21)
* **Updated** to version 8 of `babel-eslint`
* **Updated** to version 1.0.1 of `eslint-plugin-ember-standard`
* **Updated** to version 2.8.0 of `eslint-plugin-import`
* **Updated** to version 5.2.1 of `eslint-plugin-node` dependency
* **Updated** to version 1.0.2 of `eslint-plugin-ocd`
* **Updated** to version 3.6.0 of `eslint-plugin-promise`

# 7.1.1 (2017-10-05)
* **Updated** eslint-plugin-ember-standard to new version which add support for `ember-decorators` to the `computed-readonly` rule

# 7.1.0 (2017-07-06)
* **Fixed** dependencies to use versions that have appropriate `peerDependencies` settings to work with `eslint@4`. 


# 7.0.0 (2017-06-27)
* **Upgraded** `eslint-config-standard` from `5.x` to `10.x`
* **Added** explicit disabling of `import/first` rule to allow us to destructure `Ember` directly below import
* **Added** `mocha/valid-test-descriptions` (Resolves [#38](https://github.com/ciena-frost/eslint-config-frost-standard/issues/38))
* **Added** `mocha/no-mocha-arrows` (Resolves [#37](https://github.com/ciena-frost/eslint-config-frost-standard/issues/37))
* **Added** `no-var` rule to enforce using `let` or `const` instead of `var`, b/c who the heck wants to use `var` anyway ;)
* **Updated** documentation in `README.md` for additional rules we've added.
* **Upgraded** `pr-bumper` to `3.x`
* **Switched** from `npm` to `yarn` for installing dependencies.
* **Added** `node@8` to build matrix.
* **Added** the ability to detect consuming project name and add it to `localPrefixes` automatically, and tested it by locally installing in an app.

# 6.0.2 (2017-06-24)
* **Upgraded** to `eslint-plugin-ember-standard@0.0.23`

# 6.0.1 (2017-03-23)

* **Upgraded** to eslint-plugin-ember-standard 0.0.22


# 6.0.0
## Breaking Changes
* **Converted** some warning rules to error rules (the `ember` and `ocd` ones that were recently introduced)
## Non-breaking Changes
* **Updated** `pr-bumper` and `travis` settings

# 5.3.2

* **Fixed** bug with fix functionality of `ember-standard/destructure` rule where a property could end up in the variable declarator twice.


# 5.3.1

* **Fixed** bug where `Ember.String` and `Ember.Object` were being destructured when they shouldn't be.


# 5.3.0

* **Added** warning to when using [object curly spacing](http://eslint.org/docs/rules/object-curly-spacing).


# 5.2.0

* **Fixed** `ocd/sort-imports` rule to separate internal and external modules.


# 5.1.0

* **Added** plugins [ember-standard](https://github.com/ciena-blueplanet/eslint-plugin-ember-standard) and [ocd](https://github.com/ciena-blueplanet/eslint-plugin-ocd) and configured their rules as warnings (will switch to errors in the next major release).


# 5.0.1
 * **Fixed** a typo in `README.md` (thanks @juwara0 😄)



# 5.0.0
* **Added** an additional rule: `no-unused-expressions` which will disallow `expect(foo).to.be.true` style assertions. See [this issue](https://github.com/chaijs/chai/issues/726) for more info.


# 4.0.1
 * **Updated** `eslint-plugin-mocha` to allow us to stop using the fork from `job13er`



# 4.0.0
 * **New Rule** `complexity` to enforce a maximum cyclomatic complexity of 5 for all functions.

# 3.0.0
## Breaking changes
 * **Switched** to `eslint@3.x`

## New Rules

 * **Added** rule to error on a `describe.only` or `describeComponent.only` in tests
 * **Added** rule to warn on a `describe.skip` or `it.skip` (with `--fix` option 😄 )
 * **Added** rule to error on an `it('pending test')` with no test
 * **Added** rule to error on a global `it()`(not within a test suite)
 * **Added** rule to error if the `done` callback isn't called when it's defined.

# 2.1.0
* Added camelcase property for variables
* Added max-len property of 120 chars per line


tests/index-spec.js
'use strict'

const CLIEngine = require('eslint').CLIEngine
const expect = require('chai').expect
const config = require('../')

describe('config', function () {
  it('should be valid', function () {
    const cli = new CLIEngine({
      useEslintrc: false,
      configFile: 'index.js'
    })

    const code = 'let foo = 1\nlet bar = function () {}\nbar(foo)\n'
    expect(cli.executeOnText(code).errorCount).to.equal(0)
  })

  it('should extend standard', function () {
    expect(config.extends[0]).to.equal('standard')
  })

  describe('ember-standard plugin', function () {
    it('should be included', function () {
      expect(config.plugins).to.include('ember-standard')
    })

    it('should include computed-property-readonly rule', function () {
      expect(config.rules['ember-standard/computed-property-readonly']).not.to.equal(undefined)
    })

    it('should include destructure rule', function () {
      expect(config.rules['ember-standard/destructure']).not.to.equal(undefined)
    })

    it('should include import rule', function () {
      expect(config.rules['ember-standard/import']).not.to.equal(undefined)
    })

    it('should include logger rule', function () {
      expect(config.rules['ember-standard/logger']).not.to.equal(undefined)
    })

    it('should include no-set-in-computed-property rule', function () {
      expect(config.rules['ember-standard/no-set-in-computed-property']).not.to.equal(undefined)
    })

    it('should include no-settimeout rule', function () {
      expect(config.rules['ember-standard/no-settimeout']).not.to.equal(undefined)
    })

    it('should include prop-types rule', function () {
      expect(config.rules['ember-standard/prop-types']).not.to.equal(undefined)
    })

    it('should include single-destructure rule', function () {
      expect(config.rules['ember-standard/single-destructure']).not.to.equal(undefined)
    })
  })

  describe('mocha plugin', function () {
    it('should be included', function () {
      expect(config.plugins).to.include('mocha')
    })

    it('should include handle-done-callback rule', function () {
      expect(config.rules['mocha/handle-done-callback']).not.to.equal(undefined)
    })

    it('should include no-exclusive-tests rule', function () {
      expect(config.rules['mocha/no-exclusive-tests']).not.to.equal(undefined)
    })

    it('should include no-global-tests rule', function () {
      expect(config.rules['mocha/no-global-tests']).not.to.equal(undefined)
    })

    it('should include no-mocha-arrows rule', function () {
      expect(config.rules['mocha/no-mocha-arrows']).not.to.equal(undefined)
    })

    it('should include no-pending-tests rule', function () {
      expect(config.rules['mocha/no-pending-tests']).not.to.equal(undefined)
    })

    it('should include no-skipped-tests rule', function () {
      expect(config.rules['mocha/no-skipped-tests']).not.to.equal(undefined)
    })

    it('should valid-test-description rule', function () {
      expect(config.rules['mocha/valid-test-description']).not.to.equal(undefined)
    })
  })

  describe('ocd plugin', function () {
    it('should be included', function () {
      expect(config.plugins).to.include('ocd')
    })

    it('should include sort-import-declaration-specifiers rule', function () {
      expect(config.rules['ocd/sort-import-declaration-specifiers']).not.to.equal(undefined)
    })

    it('should include sort-import-declarations rule', function () {
      expect(config.rules['ocd/sort-import-declarations']).not.to.equal(undefined)
    })

    it('should include sort-variable-declarator-properties rule', function () {
      expect(config.rules['ocd/sort-variable-declarator-properties']).not.to.equal(undefined)
    })
  })

  it('should include camelcase', function () {
    expect(config.rules.camelcase).not.to.equal(undefined)
  })

  it('should include complexity', function () {
    expect(config.rules.complexity).not.to.equal(undefined)
  })

  it('should include max-len', function () {
    expect(config.rules['max-len']).not.to.equal(undefined)
  })

  it('should include no-unused-expressions', function () {
    expect(config.rules['no-unused-expressions']).not.to.equal(undefined)
  })

  it('should include no-var', function () {
    expect(config.rules['no-var']).not.to.equal(undefined)
  })

  it('should include object-curly-spacing', function () {
    expect(config.rules['object-curly-spacing']).not.to.equal(undefined)
  })

  it('should include valid-jsdoc', function () {
    expect(config.rules['valid-jsdoc']).not.to.equal(undefined)
  })
})


index.js
const fs = require('fs')
const path = require('path')

/**
 * Based on the location of this file (assumed to be node_modules/eslint-config-frost-standard/index.js)
 * we want to read in the package.json file from the root of the project, and return the project name
 * @returns {String} the package name
 */
function getProjectName () {
  const pkgJson = JSON.parse(fs.readFileSync(path.join(__dirname, '..', '..', 'package.json'), {encoding: 'utf8'}))
  return pkgJson.name
}

const localPrefixes = [
  '../',
  './',
  'dummy/'
]

try {
  const projectName = getProjectName()
  localPrefixes.push(projectName)
} catch (e) {}

module.exports = {
  extends: ['standard'],
  parser: 'babel-eslint',
  plugins: [
    'ember-standard',
    'mocha',
    'ocd'
  ],
  rules: {
    camelcase: 'error',
    complexity: ['error', 5],
    'ember-standard/computed-property-readonly': ['error', 'always'],
    'ember-standard/destructure': ['error', 'always'],
    'ember-standard/import': ['error', 'always'],
    'ember-standard/logger': ['error', 'always'],
    'ember-standard/no-set-in-computed-property': 'error',
    'ember-standard/no-settimeout': 'error',
    'ember-standard/prop-types': 'error',
    'ember-standard/single-destructure': 'error',
    'import/first': 'off', // disabling this so we can destructure Ember right after import (@job13er 2017-06-27)
    'max-len': ['error', {
      code: 120,
      tabWidth: 2
    }],
    'mocha/handle-done-callback': 'error',
    'mocha/no-exclusive-tests': 'error',
    'mocha/no-global-tests': 'error',
    'mocha/no-mocha-arrows': 'error',
    'mocha/no-pending-tests': 'error',
    'mocha/no-skipped-tests': 'warn',
    'mocha/valid-test-description': 'error',
    'no-template-curly-in-string': 'off',
    'no-unused-expressions': ['error', {
      allowTernary: true
    }],
    'no-var': 'error',
    'object-curly-spacing': ['error', 'never'],
    'ocd/sort-import-declaration-specifiers': 'error',
    'ocd/sort-import-declarations': [
      'error',
      {
        localPrefixes
      }
    ],
    'ocd/sort-variable-declarator-properties': 'error',
    'valid-jsdoc': [
      'error',
      {
        prefer: {
          'virtual': 'abstract',
          'extends': 'augments',
          'constructor': 'class',
          'const': 'constant',
          'defaultvalue': 'default',
          'desc': 'description',
          'host': 'external',
          'fileoverview': 'file',
          'overview': 'file',
          'emits': 'fires',
          'func': 'function',
          'method': 'function',
          'var': 'member',
          'arg': 'param',
          'argument': 'param',
          'return': 'returns',
          'exception': 'throws'
        },
        requireReturn: false
      }
    ]
  }
}
